be1-go/message2/jsonRPC.go

```diff
@@ -0,0 +1,79 @@
+package message2
+
+import (
+	"encoding/json"
+	"student20_pop/message2/answer"
+	"student20_pop/message2/query"
+
+	"golang.org/x/xerrors"
+)
+
+const (
+	// RPCTypeQuery ...
+	RPCTypeQuery RPCType = iota
+
+	// RPCTypeAnswer ....
+	RPCTypeAnswer
+
+	// RPCUnknown ...
+	RPCUnknown
+)
+
+// RPCType ...
+type RPCType int
+
+// JSONRPC ...
+type JSONRPC struct {
+	JSONRPC string
+
+	query.Query
+	answer.Answer
+}
```

Comment:
1. Tiny detail, but I'd call this a `JsonRpcMessage`
2. This feels semantically wrong, from an OOP perspective, a `JsonRpcMessage` does not `have` a query and/or an answer, it `is` a query `or` an answer. 

As such, I would suggest making this an interface and using a simple type switch where we need to use it. 
i.e. `type JsonRpcMessage interface {}`
That means that wherever needed, logically only in one place, we can do something along the lines of this
```
switch m := jsonRpcMsg.(type) {
    query.Query: processQuery(m)
    answer.Answer: processAnswer(m)
    default: return errors.Errorf("JsonRpcMessage is invalid")
}
```

and the Unmarshalling might look like this:
```
func UnmarshalJsonRpc(buf []byte) JsonRpcMessage, error {
    var q query.Query
    if err := json.Unmarshal(data, &q); err == nil {
        return q, nil
    }

    var a answer.Answer
    if err := json.Unmarshal(data, &a); err == nil {
        return a, nil
    }

    return nil, err
}
```

I feel it also makes the marshaling and unmarshaling logic a lot simpler and the objects contain their own fields in an obvious way.

the same logic could maybe also be applicable to `Query` vs. `Subscribe`, `Publish`, etc. but I'm not sure if that's very relevant there. I'll comment on that code directly :)

Reply:
None of those solutions are ideal. 

I definitely don't like the idea of bringing interfaces as it complexities and hides the mapping between a go struct and the json structure. The message package should provides a simple and static representation of json messages to go structs. This is why I'd like to avoid bringing interfaces and (whenever possible...) weird marshal/unmarshal functions with hidden logic. And when needed, they should be as much explicit as possible.

Your proposal to "let's try to unmarshal this message into this struct `S1`, and if it doesn't work let's try with `S2`" is fragile as two structs `SX` and `SY` could both result in a successful unmarshalling. Ok, in this case this is trivial, but...

Reply:
I agree with you on most points if we were developing a *generic solution*.

In particular, I agree with you regarding `S1` and `S2` in *generic* terms (i.e. I wouldn't write a library like this), but for the JSON-RPC protocol, it is indeed a trivial and accurate solution.

More generally, I would also argue that, for the situation at hand, interfaces make relationships more obvious, by respecting semantics (`is` vs. `has`).
There's no `JsonRpcMessage` entity in the protocol, it's an abstraction because we do need a single unmarshaling point for both query and answer. In practice they're altogether different elements.
As such, I believe with interfaces there'll actually be less weird marshaling/unmarshaling logic.

I'm tempted to say: do try and write the corresponding marshaling function to see what I mean, I'm 99% sure that the current solution will make it more complex than a trivial interface. If I'm wrong, and if the Marshal/Unmarshal logic is indeed as simple as you seem to believe, I'm happy to stick to the current solution.